### ISSUE TITLE
Branch tracing

### ISSUE THREAD

**Issue by Nukem9** [x64dbg/x64dbg#227]
This was already discussed, just adding this here to get any input.

"The IA32_DEBUGCTL MSR provides bit field controls to enable debug trace interrupts,
debug trace stores, trace messages enable, single stepping on branches, last
branch record recording..."
###  ###
* Bit 8 of DR7 represents bit 0 of IA32_DEBUGCTL. This is the LBR bit. (last branch record)
* Bit 9 of DR7 represents bit 1 of IA32_DEBUGCTL. This is the BTF bit. (single-step on branches)
###  ###
LBR in context structure (These act like registers):

```
#!

DWORD64 LastBranchToRip;
DWORD64 LastBranchFromRip;
DWORD64 LastExceptionToRip;
DWORD64 LastExceptionFromRip;
```

----------------------------------------
- Bitbucket: https://bitbucket.org/mrexodia/x64_dbg/issue/227
- Originally reported by: [Nukem9](http://bitbucket.org/Nukem9)
- Originally created at: 2014-12-24T21:36:50.728

**Comment by torusrxxx** [comment]
While branch tracing is generally available on all Intel processors, Microsoft doesn't seem to provide support on Windows. So even if it be implemented a kernel-mode driver must be used to set up that mode in the kernel. I don't want a kernel-mode driver to be bound with a user-mode software.

**Comment by Nukem9** [comment]
[Except Windows does support tracing, however undocumented](http://www.codeproject.com/Articles/517466/Last-branch-records-and-branch-tracing). DR7 bits are already set with hardware breakpoints, so this is no different.

tl;dr
```
Now, how do we access DebugCtl from usermode? It's simple, and Windows provides access to both BTF
and LBR bits of DebugCtl via bits 8 and 9 of DR7. If interested, see KiRestoreDebugRegisterState.
```
